contracts/babylon/src/contract.rs

```diff
@@ -310,15 +311,80 @@ pub fn execute(
             // TODO: Add events (#124)
             Ok(res)
         }
+        ExecuteMsg::DistributeRewards { fp_distribution } => {
+            // Check that the sender is the finality contract
+            let cfg = CONFIG.load(deps.storage)?;
+            let finality_addr = cfg
+                .btc_finality
+                .ok_or(ContractError::BtcFinalityNotSet {})?;
+            if info.sender != finality_addr {
+                return Err(ContractError::Unauthorized {});
+            }
+
+            // Check that the funds are there
+            let funds_sent_total = must_pay(&info, &cfg.denom)?;
+
+            // Calculate total rewards to send
+            let total_rewards = fp_distribution
+                .iter()
+                .map(|reward_info| reward_info.reward)
+                .sum();
+
+            // Check that the total rewards  to distribute match the funds sent
+            if funds_sent_total != total_rewards {
+                return Err(ContractError::InvalidRewards(
+                    total_rewards,
+                    funds_sent_total,
+                ));
+            }
+
+            // Get the ICS20 transfer channel ID
+            let channel_id = IBC_TRANSFER.load(deps.storage)?;
+
+            // Create memo with distribution information as JSON
+            let memo = serde_json::json!({
+                "fp_distribution": fp_distribution
+            })
+            .to_string();
```

Comment:
The memo constructed should match the structure defined in the Babylon Genesis

Reply:
Yes, taking a look at that now.

Reply:
Yes, you are right 👍